docs/development/pigs/pig-008.rst

```diff
@@ -0,0 +1,449 @@
+* Created: Jan 4th, 2018
+* Accepted:
+* Status:
+* Discussion:
+
+
+Introduction
+============
+An essential feature of Gammapy for modeling gamma-ray data will be the possibility
+of joint-likelihood analyses. This includes the joint-likelhood fitting of a model
+across multiple observations, joint-likelihood fitting of IACT and Fermi-LAT
+data, combined analysis of gamma-ray data with flux points or a combined spectral
+and cube analysis. Joint-likelihood also allows for combining different event types
+in a single analysis and timing analyses. For this reason we propose to introduce
+the abstraction layer of `Dataset` in Gammapy. A dataset bundles the reduced data
+with a parameteric model and fit statistics function. It evaluates the model and
+log-likelihood and passes it on to the fit object. Datasets can be combined by
+adding their log-likelihood values and concatenating their model parameters.
+
+
+Proposal
+========
+We propose to introduce the following classes to implement the dataset handling
+in Gammapy:
+
+
+`MapDataset`
+------------
+To enable the standard combined spectral and spatial analysis we propose to
+introduce a `MapDataset` class. A `MapDataset` bundles the counts data, source model,
+IRFs, background model, corresponding to a given event selection.
+
+It is supposed to work as following:
+
+.. code ::
+
+    from gammapy.cube import MapDataset
+
+    model = SourceModel.read("model.yaml")
+    background = BackgroundModel.read("stacked/background.fits")
+
+    counts = Map.read("stacked/counts.fits")
+    exposure = Map.read("stacked/exposure.fits")
+    edisp = Map.read("stacked/edisp.fits")
+    psf = Map.read("stacked/psf.fits")
+
+    dataset = MapDataset(
+        counts=counts,
+        exposure=exposure,
+        edisp=edisp,
+        psf=psf,
+        mask=mask,
+        model=model,
+        background=background_model,
+        likelihood="cash"
+        )
+
+    fit = Fit(dataset)
+    fit.optimize()
+
+The edisp and psf maps can be defined on a coarse spatial geometry. The counts,
+exposure and background map are supposed to be on the same spatial geometry,
+chosen by the user for the analysis.
+
+The `MapDataset` implements at least the following methods:
+
+.. code::
+
+    dataset = MapDataset()
+
+    # the map dataset might require some precomputations
+    dataset.setup()?
+
+    dataset.npred()
+    dataset.likelihood_per_bin()
+    dataset.likelihood(mask)
+
+The `.setup()` method creates cutouts of the exposure map for the individual
+model components, assigns the correspoding IRFs to the model component and
+bundles all into `SourceIRFModel` or `ModelEvaluator()` classes. The list of model
+evaluators is cached on the `MapDataset` and used later to efficiently compute
+the predicted number of counts.
+
+The `mask` argument to the `.likelihood` method allows to set a global fitting
+mask, when mutiple `MapDataset` are fitted jointly.
+
+
+`MapDatasetOnOff`
```

Comment:
Are you sure a separate class for on/off is a good idea?

I think the options are:
- separate class
- subclass on/off from on
- one class with optional data members for off (in HESS we have this as BgMaps and it works)

All of those have pros / cons in terms of simplicity vs code duplication.

Suggest to change in the PIG to put 2-3 sentences mentioning the options, without deciding on one.
But if @registerrier and you have discussed this and are sure you want separate classes for on/off, OK to leave as-is.

Reply:
Whether to introduce separate `...OnOff` datasets is probably something we should discuss tomorrow...